File: src/2022-intern/1582.special-positions-in-a-binary-matrix.js
/*
 * @lc app=leetcode id=1582 lang=javascript
 *
 * [1582] Special Positions in a Binary Matrix
 */

// @lc code=start
/**
 * @param {number[][]} mat
 * @return {number}
 */
var numSpecial = function (mat) {
  const m = mat.length;
  const n = mat[0].length;

  /**
   * 在某一行里，有哪些列是 1
   */
  const rowToOneCols = new Array(m);
  for (let i = 0; i < m; i++) {
    rowToOneCols[i] = [];
  }

  /**
   * 在某一列里，有几行是 1
   */
  const colToOneRowCount = new Array(n);
  colToOneRowCount.fill(0);

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (mat[i][j] === 1) {
        rowToOneCols[i].push(j);
        colToOneRowCount[j]++;
      }
    }
  }

  let result = 0;

  for (let i = 0; i < m; i++) {
    if (rowToOneCols[i].length === 1) {
      const col = rowToOneCols[i][0];
      if (colToOneRowCount[col] === 1) {
        result++;
      }
    }
  }

  return result;
};
// @lc code=end

File: src/2026/1582.special-positions-in-a-binary-matrix.cpp
/*
 * @lc app=leetcode id=1582 lang=cpp
 *
 * [1582] Special Positions in a Binary Matrix
 */

#include <vector>

// @lc code=start
class Solution {
 public:
  int numSpecial(const std::vector<std::vector<int>>& mat) {
    const int M = mat.size();
    const int N = mat[0].size();
    std::vector<int> row_one_count(M);
    std::vector<int> col_one_count(N);

    for (int i = 0; i < M; i++) {
      for (int j = 0; j < N; j++) {
        row_one_count[i] += mat[i][j];
        col_one_count[j] += mat[i][j];
      }
    }

    int result = 0;

    for (int i = 0; i < M; i++) {
      for (int j = 0; j < N; j++) {
        result +=
            mat[i][j] == 1 && row_one_count[i] == 1 && col_one_count[j] == 1;
      }
    }

    return result;
  }
};
// @lc code=end
